### DISCUSSION TITLE
small sound clip playing double in iOS but it's ok in android

### DISCUSSION THREAD

**Discussion by meehjeevan** [axmolengine/axmol#2216] [Q&A]
I have used a button click sound clip and it's playing double in iOS but it's ok in android. 
I have played the sound with this code: 
`AudioEngine::play2d("sounds/button6.mp3");`
I couldn't upload the audio file here.. 
The sound clip length is 0.14 seconds only.
How do I solve it ? Any idea?

**Comment by aismann** [comment]
> How do I solve it ? Any idea?

Sure. Start a debugging session.

**Reply by meehjeevan** [reply]
The function which plays the sound is called once... no double call. 
Only it sounds double when its on iOS devices. I tried putting silence at the end of the sound clip too. But still no luck.

**Reply by rh101** [reply]
@meehjeevan If you haven't already done so, create a minimal test project, add the `AudioEngine::play2d("sounds/button6.mp3");` call in an event handler (button click or similar), and check if the issue still occurs.  If it does happen, then change it to a different sound file, and test it again.  

If you're 100% sure it is happening in that minimal project, then please upload the project with whatever audio file you can share (preferably to a github repo).  If you can't share any of your audio files, then find an open source one that works to showcase the problem, and use that for the test project.

**Reply by meehjeevan** [reply]
ok, I have created a project for testing sounds only. I have put same sound with 4 different formats. All format plays double in iOS.
Please check:
https://github.com/meehjeevan/soundTest

**Reply by rh101** [reply]
> ok, I have created a project for testing sounds only. I have put same sound with 4 different formats. All format plays double in iOS. Please check: https://github.com/meehjeevan/soundTest

Thanks, the project was very helpful in reproducing the issue.  A few things regarding this problem:

1. I think that the audio file isn't actually playing twice, but it's more like the audio device is stuck playing the same small section of audio for a certain period of time (like it's frozen).  @halx99 Could this have something to do with locking or some kind of delay in the code?
2. If you add approximately 25-50ms of silence to the start of the audio, you will notice that the file plays normally, without that strange repeating sound at the start.  This may actually confirm that it's only the start of the file that seems to have that distortion due to the audio device.  
3. You do not need the silence at the end of the audio clip, so it would be best to remove it. 

The quick fix is to add a small amount of silence at the start of the audio clip; around 25milliseconds worth of silence seemed to work fine with my tests, so start with that amount.  

The only reason you noticed this issue is because the audible section of the "Button" sound track lasts barely 20milliseconds, which is around as long as the audio device seems to have the issue.  It's not something anyone would notice when playing tracks with audible sections lasting longer than that.

The audio playback issue did not happen on Windows either, so it is likely to be specific to iOS.

**Reply by meehjeevan** [reply]
ok, I added 30 ms silence at start and now it's fine. thanks.
